<a href="https://colab.research.google.com/github/kareemrb27/GenAIIK/blob/main/Agentic_AI_Workshop.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### What is Agentic AI?
Agentic AI refers to AI systems designed with autonomous decision-making capabilities, enabling them to perceive, reason, and take actions dynamically. These systems leverage multi-agent architectures,and LLMs to perform complex tasks with minimal human intervention. They are widely used in automation, cybersecurity, finance, and enterprise AI for intelligent workflows. Tools like LangChain, CrewAI, AutoGen, and LlamaIndex facilitate building and orchestrating these AI agents.

<center><img src="https://cdn.prod.website-files.com/66c435ec15ed715aec9ee3fe/66fd1383a0e470c48179a973_66fd136aff11aa6797621e82_What%2520is%2520agentic%2520AI.jpeg" width=600/></center>

### How is it different from Generative AI?

<center><img src="https://cdn.prod.website-files.com/66c435ec15ed715aec9ee3fe/66e78a95f15da5104e75ac08_66e78a880dcb8cb0f1229412_Agentic%2520AI%2520vs%2520Generative%2520AI.png" width =600/></center>

## Basics of RAG

Before we start coding, lets go over a few questions and get it clarified.

### 1. What is RAG?
Imagine you're writing an article about climate change, but instead of relying only on what you remember, you search online for recent studies and data to support your writing. RAG works the same way—it combines a powerful AI language model with a search system that retrieves relevant information from an external knowledge source, like a database or documents. This helps generate more accurate and informative responses.

### 2. Why is RAG important?
RAG is crucial because AI models, like ChatGPT, can sometimes "hallucinate" or provide outdated or incorrect information. By retrieving facts from trusted sources before generating responses, RAG ensures the answers are more reliable, up-to-date, and contextually relevant.

### 3. What’s the difference between RAG and a standard AI chatbot?
A standard chatbot relies only on pre-trained knowledge, which may be limited or outdated. RAG-enhanced chatbots, however, actively retrieve fresh, relevant information from external sources, ensuring better accuracy and up-to-date insights.

### 4. What are the key components of RAG?
RAG consists of two main parts:

- Retriever: Finds the most relevant documents or data from a knowledge base (e.g., a search engine or database).
- Generator: Uses the retrieved information to produce a coherent and accurate response.

### 5. Usecases in real-life
- Customer Support: AI chatbots retrieve knowledge base articles to provide better responses to customer queries.
- Healthcare: Doctors can get AI-assisted summaries of patient records and the latest medical research.
- Legal Services: Lawyers can search through legal documents and case studies to build stronger cases.

### Understanding the Limitation of the LLM

In [ ]:
# Importing the OpenAI library to interact with OpenAI's API services.
from openai import OpenAI

In [ ]:
import os  # Importing the os module to interact with environment variables
import getpass  # Importing getpass to securely input sensitive information

# Prompting the user to securely enter their OpenAI API key without displaying it on the screen
OPENAI_API_KEY = getpass.getpass("Enter your OpenAI API key: ")

Enter your OpenAI API key: ··········


In [ ]:
# Creating an instance of the OpenAI client using the provided API key.
client = OpenAI(api_key=OPENAI_API_KEY)

In [ ]:
# Defining the prompt to query the LLM
prompt = ''' What was uber's revenue in 2022? '''

In [ ]:
# Sending a request to the OpenAI API to generate a chat response
openai_response = client.chat.completions.create(
    model='gpt-3.5-turbo',  # Specifying the model to use;
    # Note: An older model chosen for testing purposes because the cutoff is 2021 whereas prompt is querying details about 2022
    messages=[{'role': 'user', 'content': prompt}]  # Creating a structured message for the AI model
)


In the above code, while creating a structured message for the AI model,  `role `defines the speaker (user input) and `content` contains the actual query stored in the `prompt` variable.

We are structuring the input this way because OpenAI's chat models require a specific format to understand and process conversations effectively. Assigning roles like 'user' helps the AI distinguish between different participants in the conversation, ensuring it provides relevant and context-aware responses.

In [ ]:
# Accessing the generated response from the AI model.
openai_response.choices[0].message.content
# Note:'choices' contains multiple response options, we take the first one ([0]),
# 'message' holds the response details, and 'content' extracts the actual text generated by the AI.

"I cannot provide real-time data as I am an AI programmed with information up to September 2021. I recommend checking Uber's official website or financial reports for the most accurate and up-to-date revenue information for 2022."

### Interpretation:
Why is the LLM not able to answer the query?

`gpt-3.5-turbo` does not have access to data after 2021 because of its cutoff.

#### Do it yourself:
Try changing the model to `gpt-4o-mini` and observe how the output changes!

As you can see above, the LLM we used(gpt 3.5) doesn't have access to the latest data. Now as LLMs get updated, the training cut-off date may or may not have access to more information. However, it's always a good idea to understand how to improve the context of our prompt

### Making the LLM context-aware

Next Step: Let's check Uber's [financial report ](https://s23.q4cdn.com/407969754/files/doc_events/2024/May/06/2023-annual-report.pdf)

On Page 54, of the above document it states:

"Revenue was 37.3 billion, up 17% year-over-year. Mobility revenue increased 5.8 billion primarily attributable to an increase in
Mobility Gross Bookings of......"

In [ ]:
## Let's create the above context for the prompt
# Defining a context string with revenue details retrieved from an external source.
retrieved_context = '''Revenue was $37.3 billion, up 17% year-over-year. Mobility revenue increased $5.8 billion primarily attributable to an increase in
               Mobility Gross Bookings of 31% year-over-year.'''

In [ ]:
## Let's modify our prompt now
# Creating a prompt by embedding the retrieved context into a question for the AI model.

prompt = f"What was Uber's revenue in 2022? Check in {retrieved_context}"

# Note: The AI is being asked to analyze the given context and provide Uber's revenue for 2022

In [ ]:
## Let's ask the LLM again
openai_response = client.chat.completions.create(
    model = 'gpt-3.5-turbo',
    messages = [{'role': 'user', 'content': prompt}])

In [ ]:
# Accessing the generated response from the AI model.
openai_response.choices[0].message.content

"Uber's revenue in 2022 was $37.3 billion."

### Interpretation:
How is the LLM able to answer the same question now?

The LLM can now answer the question accurately because the relevant financial data is explicitly provided in the `retrieved_context`, allowing the model to reference it directly instead of relying on its pre-trained knowledge.

As you saw in the example above, we

- **retrieved** the context from an external source
- **augmented** our prompt that passes to the LLM, and
- **generated** the response

This is Retrieval Augmented Generation in a nutshell!

### Basic RAG app architecture

In the previous example, we ***manually*** retrieved the context from the given file which for all purposes is impractical!

Therefore, we have to devise a strategy that enables us to:

- Take the query from the user
- Identify the documents from the external source that might be relevant for the query.
- Pass those documents' information as context to the LLM
- LLM then generates the final response

To do the above, we can follow a simple standard architecture as shown below (Image source - https://huyenchip.com/2024/07/25/genai-platform.html)

<center><img src="https://huyenchip.com/assets/pics/genai-platform/3-rag.png" width=500 height=400/></center>

As you can see in the above image, the retriever would be the key component of this entire architecture.

To build the retriever, we have to follow these steps:

- Connect to the document source
- Break the documents down to manageable chunks.
   - This is due to the fact that taking in the entire document source for building the context will exceed the token limits of the LLM.
   - This process is also called **Chunking**.
- Perform a search for the most relevant chunks based on the given query.
- Pass those relevant chunks to the LLM.

For performing the search or retrieval process, we will be following an **embedding-based approach.**

<center><img src="https://cdn.prod.website-files.com/640248e1fd70b63c09bd3d09/653fd23f1565c0c1da063efc_Semantic%20Search%20Text%20Embeddings%20(1).png" width =500/></center>

### Understanding Embedding based approach

In the embedding based approach:

- We convert the document chunks in the database to vector embeddings and store it in a vector store.

- Convert the given user query to an embedding.

- Find the document chunks whose vector embeddings are closest to the given query embedding using a vector search algorithm like FAISS (Facebook AI Similarity Search)

<center><img src="https://miro.medium.com/v2/resize:fit:1400/1*h_btyitJX79d-gFE8RaMQg.png" width=500/></center>

### Tools for building the RAG App

Now that we are familiar with the overall architecture, we can now go ahead and structure the tools that we'll use for the upcoming demonstration:

- OpenAI LLM (model - GPT 4o-mini): This will be our primary model for generating the responses
- LangChain: Langchain is a powerful framework for orchestrating different layers in the RAG app. We shall use this to build the retriever end-to-end and also connect with other tools for tasks such as
    - Chunking - RecursiveCharacterTextSplitter
    - Embedding Model - OpenAIEmbeddings
    - Vector Search Model - FAISS
- Gradio: This will help in building a simple UI at the end.

## Problem Statement
Shopping online can be overwhelming. You search for a simple pair of shoes, but end up scrolling through countless options—many irrelevant, some too expensive, others just not right. Traditional search engines rely on keywords, often missing what you truly need.

Let's build an AI-powered product discovery chatbot changes this. Using advanced language models and vector-based search, it goes beyond keywords to understand your intent, offering personalized, context-aware recommendations in seconds.

<center><img src="https://www.pranathiss.com/static/assets/images/ai-powered-chatBot.webp" width=500/></center>

This smart solution enhances the shopping experience, increasing customer satisfaction, engagement, and conversions. The future of e-commerce is here—smarter, intuitive, and built for you.

### Dataset Used:
The given dataset contains information about various products, including their IDs, descriptions, and specifications. Below is a detailed description of each column and the type of data it contains.

You can download the entire dataset [here](https://www.kaggle.com/datasets/piyushjain16/amazon-product-data).
Or you can download the smaller sample dataset [here](https://drive.google.com/file/d/1ohd9xo19HmDVIwpXPf_IyMkwr29gJJxR/view?usp=drive_link).

#### Column Descriptions:
- `PRODUCT_ID (Integer)`
A unique identifier assigned to each product.
Example: 1925202, 2673191
- `TITLE (String)`
The name or title of the product, usually a brief summary.
Example: "ArtzFolio Tulip Flowers Blackout Curtain for D...",
"Marks & Spencer Girls' Pyjama Sets T86_2561C_N..."
- `BULLET_POINTS (List of Strings / NaN)`
A list of key product features and benefits in bullet format.
- `DESCRIPTION (String / NaN)`
A detailed textual description of the product, including specifications, features, and usage instructions.
Example: "Specifications: Color: Red, Material: Aluminium..."
- `PRODUCT_TYPE_ID (Integer / NaN)`
A numeric identifier indicating the type or category of the product.
Example: 1650, 2996, 7537
- `PRODUCT_LENGTH (Float)`
The length of the product, likely measured in millimeters or inches.
Example: 2125.98, 393.7, 748.031495

### Steps:

1. **Data Preparation**:
   - Load and process the dataset using pandas

2. **Vector Store Setup**:
   - Convert product descriptions into embeddings.
   - Store embeddings in a vector database.

3. **Building the Chatbot**:
   - Use LangChain to create an LLM pipeline.
   - Develop a simple chatbot to answer product-related queries.

4. **Creating a UI**:
   - Implement a Gradio-based UI for user interaction.

In [ ]:
# Installing the LangChain Hub package to access and manage pre-built AI chains, prompts, and agents.
!pip install langchainhub

# Installing the LangChain OpenAI integration to use OpenAI models within LangChain workflows.
!pip install langchain-openai

# Installing the core LangChain library for building LLM-based applications, including chaining, memory, and retrieval capabilities.
!pip install langchain

# Installing the community version of LangChain, which includes integrations and tools contributed by the community.
!pip install langchain-community

# Installing FAISS (Facebook AI Similarity Search) for efficient similarity-based search on text embeddings.
!pip install faiss-cpu

# Installing Gradio, a framework to create web-based UIs for AI models and applications easily.
!pip install gradio


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 54.4/54.4 kB 4.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 46.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 66.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.8/50.8 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 27.5/27.5 MB 42.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 MB 12.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 321.8/321.8 kB 22.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 94.8/94.8 kB 7.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 98.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.5/71.5 kB 6.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.3/62.3 kB 5.4 MB/s eta 0:00:00
  Attempting uninstall: markupsafe
    Found existing installation: MarkupSafe 3.0.2
    Uninstalling MarkupSafe-3.0.2:
      Successfully

In [ ]:
# Importing the KaggleHub library to interact with datasets and models available on Kaggle.
import kagglehub

# Importing the CSV module for reading and writing CSV files.
import csv

# Importing pandas for data manipulation and analysis.
import pandas as pd

# Importing numpy for numerical operations and handling arrays efficiently.
import numpy as np

# Importing os to interact with the operating system, such as environment variables and file paths.
import os

# Importing getpass to securely handle user input (e.g., API keys or passwords).
import getpass

import math

### STEP 1: Data Preparation

In [ ]:
# Mount Google Drive
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


In [ ]:
# Loading the data
# df = pd.read_csv("/content/gdrive/MyDrive/datasets/sample_dataset.csv",index_col=0)
df = pd.read_csv("/content/gdrive/MyDrive/IK/project_based/sample_dataset.csv", index_col=0)

In [ ]:
print(df.to_csv(columns=['DESCRIPTION'], sep='\t', index=False))

DESCRIPTION
""
""
Specifications: Color: Red, Material: Aluminium, Voltage: 12V, dB: 130 dB (around), Material: Aluminum Pump Head + Steel Pump Body + ABS Shell and Parts DB output: 130db Voltage: 12v Sound Type: Dual Tone Application: 12V Voltage Vehicles With Battery Above 20A Package included: 1 x Dual Tone Air Horn Compatible With SX4
AISHAH Women's Lycra Cotton Ankel Leggings. Brand: ALISHAH Length: Ankel Length Leggings. Size Chart Medium: 28 - 30 inch Large: 30 - 32 inch X-Large: 32 - 34 inch 2X-Large: 34 - 38 inch The color fade will not be there in washing. This cotton stretch leggings is designed to provide absolute comfort and body fit. Pair it with short or long kurta for modern look. High on style, fit and finish, this leggings is sure to lend you a sophisticated look. Please wash the dark colors separately in order to improve the product life. Disclaimer: Product color may slightly vary due to photographic lighting sources or your monitor settings. Please order according 

**Constructing the text data**

It's useful to use both `Title` and `Description`. To help downstream models understand which content is title and which content is description, we will add a prefix explaining which section is title and which is description. So each row should look like

```
Title
{Title}
Description
{Description}
```

In [ ]:
## Let's construct the text data
# Initializing empty lists to store product descriptions and their lengths
product_description = []
product_description_len = []

# Iterating through each row in the dataframe df2
for row in df.iterrows():
    product = ""  # Initialize an empty string to accumulate product details

    # Extracting the product title from the current row
    title = row[1]["TITLE"]

    # Checking if the title is valid (not NaN or missing)
    if type(title) != float or not math.isnan(title):
        product += "Title\n" + title + "\n"  # Append the title to the product description

    # Extracting the product description from the current row
    description = row[1]["DESCRIPTION"]

    # Checking if the description is valid (not NaN or missing)
    if type(description) != float or not math.isnan(description):
        product += "Description\n" + description + "\n"  # Append the description to the product details

    # Check if either title or description was added
    added_content = title or description
    if added_content:
        product = product.strip()  # Remove any leading/trailing whitespace
        product_description.append(product)  # Add the formatted product details to the list
        product_description_len.append(len(product))  # Store the length of the product description


In [ ]:
# Checking the length of the data
print(f"Number of elements {len(product_description)}")

Number of elements 100


In [ ]:
# Check a sample product description data
print(product_description[2])

Title
PRIKNIK Horn Red Electric Air Horn Compressor Interior Dual Tone Trumpet Loud Compatible with SX4
Description
Specifications: Color: Red, Material: Aluminium, Voltage: 12V, dB: 130 dB (around), Material: Aluminum Pump Head + Steel Pump Body + ABS Shell and Parts DB output: 130db Voltage: 12v Sound Type: Dual Tone Application: 12V Voltage Vehicles With Battery Above 20A Package included: 1 x Dual Tone Air Horn Compatible With SX4


In [ ]:
# Print the total number of product descriptions processed
print("Number of items", len(product_description_len))

# Print the minimum length of the product descriptions
print("Min Length of the description:",np.min(product_description_len))

# Print the average (mean) length of the product descriptions
print("Avg Length of the description:",np.mean(product_description_len))

# Print the maximum length of the product descriptions
print("Max Length of the description:",np.max(product_description_len))

Number of items 100
Min Length of the description: 18
Avg Length of the description: 385.9
Max Length of the description: 1834


### Interpretation:

What does the above result signify about the data?


*   A minimum length of 18 suggests that some product descriptions might be too brief
*  With an average length of 385.9 characters, most product descriptions contain a reasonable amount of information



### STEP 2: Vector Store Setup

Let's try to get a few of the basic questions answered about vector stores before we start using it.

### What is a vector store?
A vector store is a specialized database that stores data in the form of numerical vectors, allowing efficient searching and retrieval based on similarity rather than exact matches.

### Why do we need a vector store?
Traditional databases rely on exact keyword matches, which can miss relevant information. A vector store helps find similar content by understanding relationships and meaning in data.

### How does a vector store work?
It converts text, images, or other data into numerical vectors using AI models, then stores these vectors and retrieves similar ones using techniques like cosine similarity.

### How does a vector store improve search results?
It enables searches based on meaning rather than just keywords, providing more relevant results even if the exact terms don't match.

### What are some popular vector store tools?
- FAISS (Facebook AI Similarity Search)
- Pinecone
- Weaviate
- Chroma

### What is an embedding, and how does it relate to a vector store?
An embedding is a numerical representation of data (e.g., text, image) that captures its meaning. These embeddings are stored in a vector store for efficient retrieval.


Our next step is
-  to convert the `product_description` to chunks
-  convert each chunk to embedding
-  store it in vector store for searching

As discussed earlier we shall use `LangChain` to perform these steps.

LangChain is a framework that helps developers build applications powered by large language models (LLMs) like GPT by providing tools for various tasks to be carried out like retrieving relevant information from databases, etc.

In [ ]:
# Importing RecursiveCharacterTextSplitter from LangChain for chunking large
# text into smaller, manageable pieces.
# This helps in optimizing text for processing and retrieval.
from langchain_text_splitters import RecursiveCharacterTextSplitter

# Importing OpenAIEmbeddings from LangChain to generate numerical vector
# representations (embeddings) of text.
# These embeddings capture the semantic meaning of the text for efficient
# similarity searches.
from langchain_openai import OpenAIEmbeddings

# Importing FAISS (Facebook AI Similarity Search) from LangChain's
# community package.
# FAISS is used for storing and retrieving embeddings efficiently by finding
# similar vectors.
from langchain_community.vectorstores import FAISS


In [ ]:
# Setting the OpenAI API key as an environment variable.
os.environ["OPENAI_API_KEY"] = OPENAI_API_KEY

In [ ]:
# Split the input text using Recursive Character Chunking
# See this for more details https://python.langchain.com/v0.1/docs/modules/data_connection/document_transformers/recursive_text_splitter/

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=250,
    chunk_overlap=20,
    length_function=len,
    is_separator_regex=False,
)
documents = text_splitter.create_documents(product_description)

### Code Explanation:
The above code initializes a `RecursiveCharacterTextSplitter` to break down product_description into smaller text chunks of 250 characters each, with a 20-character overlap to preserve context between chunks. The `create_documents` function processes the text list and generates structured document chunks for efficient retrieval and analysis.

### Why do we need overlap?
Overlap is needed to ensure continuity and preserve context between chunks, preventing important information from being cut off at chunk boundaries.

This helps LLMs better understand the text when processing each chunk independently, improving retrieval accuracy and response quality.

In [ ]:
# Create an embedding model using LangChain.
# One option is using https://python.langchain.com/docs/integrations/text_embedding/openai/
# See https://python.langchain.com/docs/integrations/text_embedding/ for a list of available embedding models on LangChain
embeddings = OpenAIEmbeddings()

In [ ]:
# Create a vector store using the created chunks and the embeddings model
vector = FAISS.from_documents(documents, embeddings)

### What have we done so far?
1. Data Preparation: Extracted the product description data
2. Data Chunking: Converted the entire data into multiple manageable chunks
3. Chunks to Embeddings: Converted the broken down chunks into embeddings
4. Storage in a Vector DB: Stored the resulting embeddings of chunks in a vector store for effective retrieval.


### What is remaining?
- Building the chatbot
- Building the Gradio UI

### STEP 3: Building the chatbot

Now that we have converted the documents to embeddings, our next step is to
- build a retriever that uses the vector store to retrieve the documents
- create a prompt template that contains the augmented context using the retrieved documents

In [ ]:
# Importing ChatOpenAI from LangChain to interact with OpenAI's language models,
# such as GPT, for generating responses.
from langchain_openai import ChatOpenAI

# Importing ChatPromptTemplate to create structured prompts for the chatbot,
# ensuring consistent interactions with the AI model.
from langchain_core.prompts import ChatPromptTemplate

# Importing OpenAIEmbeddings to convert text data into numerical vector
# representations for similarity search and retrieval.
from langchain_openai import OpenAIEmbeddings

# Importing create_stuff_documents_chain to combine and process retrieved
# documents for meaningful AI-generated responses.
from langchain.chains.combine_documents import create_stuff_documents_chain

# Importing create_retrieval_chain to build a chain that retrieves relevant
# documents from a vector store and generates AI responses.
from langchain.chains import create_retrieval_chain

# Importing StrOutputParser from LangChain to parse the output
from langchain_core.output_parsers import StrOutputParser

#### Code Explanation:
- `ChatOpenAI` – Used to access OpenAI models for chatbot functionality.
- `ChatPromptTemplate` – Helps structure queries to ensure better responses.
- `OpenAIEmbeddings` – Converts text into vector form for similarity-based retrieval.
- `create_stuff_documents_chain` – Combines retrieved documents meaningfully before passing to the LLM.
- `create_retrieval_chain` – Automates the process of retrieving and utilizing relevant content for AI responses.
- `StrOutputParser` - For processing the output of language models, ensuring that the output is returned as a plain string

In [ ]:
# Initializing the ChatOpenAI model to interact with OpenAI's GPT model.
llm = ChatOpenAI(api_key=os.environ["OPENAI_API_KEY"], model = 'gpt-4o-mini')

In [ ]:
# Importing the output parser to process and format the model's response into a readable string format.
output_parser = StrOutputParser()

# Creating a prompt template that instructs the LLM to act as a
# customer service agent.
# The prompt takes two parameters:
#   1. {context} - Relevant information retrieved from the document store.
#   2. {input} - The user's question.
# The model is instructed to base its answer solely on the provided context.
prompt = ChatPromptTemplate.from_template(
    """Answer the following question based only on the provided context:

    <context>
    {context}
    </context>

    Question: {input}""",
    output_parser=output_parser                  # The output parser ensures that the response is returned in a structured string format.
)

# Creating a document processing chain using the LLM and the defined prompt
# template.
# This chain takes a list of retrieved documents and passes them as context to
# the model for generating responses.
document_chain = create_stuff_documents_chain(llm, prompt)

# Alternative chain creation method:
# Using the "|" (pipe) operator to link the prompt with the language model (llm),
# meaning the input first goes to the prompt and then to the model for
# response generation.
# document_chain = prompt | llm


#### Code Explanation:
- A structured prompt is created using `ChatPromptTemplate` to guide the LLM in answering questions based solely on provided context.
- The prompt includes placeholders `{context}` and `{input}` to dynamically inject relevant information.
- `StrOutputParser()` ensures that the LLM's response is formatted as plain text for easy processing and display.
- `create_stuff_documents_chain(llm, prompt)` combines the language model (LLM) with the prompt to form a processing chain. This chain takes retrieved documents as input and generates AI-driven responses.
- Alternate way:  `prompt | llm` is a more concise way to chain the prompt and model, achieving the same functionality with a cleaner syntax.

In [ ]:
# Create a retriever from the vector store for fetching relevant documents
# See https://python.langchain.com/v0.1/docs/modules/data_connection/retrievers/vectorstore/
retriever = vector.as_retriever()

# Create a retrieval chain that first retrieves relevant documents and then processes them using the document chain
retrieval_chain = create_retrieval_chain(retriever, document_chain)


#### Code Explanation:
- The `vector.as_retriever()` converts the vector store into a retriever to find documents based on query similarity.
- The `create_retrieval_chain()` connects the retriever with the document processing pipeline, ensuring the AI receives relevant context before generating responses.

This setup enables the AI to provide accurate answers by first retrieving and then processing relevant documents.

In [ ]:
# Invoking the retrieval chain to process the user's query.
# The query "what are some of the best shoes available?" is passed as input.
# The retrieval chain first fetches relevant product descriptions from the vector store,
# then processes them using the document chain to generate a meaningful LLM response.
retrieval_chain.invoke({"input": "what are some of the best shoes available?"})

{'input': 'what are some of the best shoes available?',
 'context': [Document(id='030d7be1-3321-4423-859b-73eb92d50838', metadata={}, page_content="Title\nadidas Men's Predator 18+ FG Firm Ground Soccer Cleats\nDescription\nadidas Predator 18+ FG- Black 7.5"),
  Document(id='e98e3568-1bb6-41ed-b295-5376c909e042', metadata={}, page_content="Title\nPUMA Cali Sport Clean Women's Sneakers White Leather (37540701)"),
  Document(id='ddf05f60-bc73-4e05-9c84-0f793358146e', metadata={}, page_content="Title\nKenneth Cole REACTION Men's Crespo Loafer B Shoe, Cognac, 10 M US"),
  Document(id='772c1c4b-2a66-4f39-ba95-02c0ecaa0cef', metadata={}, page_content="The Remora Climbing Shoe is Mad Rock's do-it-all slipper for climbers who can't have separate shoes for boulders, sport routes, and gyms. With a moderately stiff, slightly downturned design, the Remora performs on any climb at steep to vertical")],
 'answer': "Based on the provided context, some of the best shoes available include:\n\n1. adidas

In [ ]:
# Fetching the final answer from the retrieval chain by invoking it with a user query.
# The ['answer'] key extracts the final LLM-generated answer from the response dictionary.
resp = retrieval_chain.invoke({"input": "what are some of the best shoes available?"})['answer']
print(resp)

The context provided mentions several specific shoe models, each suited for different activities. Some of the best shoes available based on the context include:

1. **adidas Men's Predator 18+ FG Firm Ground Soccer Cleats** - Designed for soccer on firm ground surfaces.
2. **PUMA Cali Sport Clean Women's Sneakers** - A stylish white leather sneaker for everyday wear.
3. **Kenneth Cole REACTION Men's Crespo Loafer B Shoe** - A cognac-colored loafer suitable for casual or semi-formal occasions.
4. **Mad Rock Remora Climbing Shoe** - A versatile climbing shoe for various climbing activities, suitable for both bouldering and sport routes.

These options cater to athletic, casual, and specialized needs.


Now, we got the answer! But, the formatting is not very good, right? Lets create a simple UI for our bot.

### STEP 4: Building a simple Gradio UI

Gradio is an open-source Python library that makes it easy to build interactive user interfaces for machine learning models, APIs, and data science workflows. It allows developers to create shareable web-based UIs with just a few lines of code.

To build the gradio app we'll utilize the following steps:

- Modularize the entire RAG pipeline using a single function
- Create the building blocks for the UI.
- Connect the UI with the function

In [ ]:
# Function to process the user query and return formatted product names
def final_response(user_query):
    # Invoking the retrieval chain with the user's query to fetch relevant product information
    response = retrieval_chain.invoke({"input": user_query})['answer']

    # Creating a prompt to instruct the LLM to format the response properly
    # The prompt asks the LLM to extract only product names from the
    # retrieved response
    prompt = f"Format the responses properly in {response}. Just return the product names, no other text"

    # Sending the formatted prompt to the GPT-4o-mini model for processing
    openai_response = client.chat.completions.create(
        model='gpt-4o-mini',  # Using GPT-4o-mini model for response generation
        messages=[{'role': 'user', 'content': prompt}]  # Providing the prompt to the model
    )

    # Extracting and returning the LLM-generated response containing only the
    # product names
    return openai_response.choices[0].message.content


In [ ]:
# Printing the final response
print(final_response("what are some of the best shoes available?"))

1. adidas Men's Predator 18+ FG Firm Ground Soccer Cleats  
2. PUMA Cali Sport Clean Women's Sneakers (White Leather)  
3. Kenneth Cole REACTION Men's Crespo Loafer B Shoe (Cognac)  
4. Mad Rock Remora Climbing Shoe  


In [ ]:
# Importing the Gradio library to create a simple web-based user interface
import gradio as gr

# Creating the Gradio interface for the product recommendation system
app = gr.Interface(
    fn=final_response,        # The function that processes user input and returns recommendations
    inputs="text",            # Input component: a text box for users to enter their query
    outputs="text",           # Output component: a text box to display the AI-generated response
    title="Review Genie",     # The title of the web interface
    description="Type your question below to get the recommendations",# A brief description displayed to users
    theme="Ocean",
    allow_flagging="never"    # Disabling the flagging feature to remove the "Flag" button
)

# Launching the Gradio app to start the interface and make it accessible via web browser
app.launch()


/usr/local/lib/python3.11/dist-packages/gradio/interface.py:403: UserWarning: The `allow_flagging` parameter in `Interface` is deprecated.Use `flagging_mode` instead.
  warnings.warn(


Running Gradio in a Colab notebook requires sharing enabled. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://27dbcbd1f342dee386.gradio.live

This share link expires in 72 hours. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


### What are the limitations in this?
### Limitations of the LLMs

Large Language Models (LLMs) like ChatGPT, GPT-4, and other AI systems are incredibly powerful, but they have a key limitation: they rely on pre-trained knowledge, which becomes outdated over time.

Earlier we saw how Retrieval Augmented Generation (RAG) solves this problem by providing access to external knowledge sources and improve the responses from the LLMs. However, these external knowledge sources are static resources and can also get outdated after a period of time.

Take the following example:

`"What was the Stock Price average of Uber in the last week?"`

If we were to design an AI system using RAG for answering questions as above, it would require a constant update of the knowledge base every week. This can be an expensive and inefficient process. A much better way to implement this would be to use **Agentic Tools** that can retrieve real-time information.

This is where ***Tavily*** plays a crucial role by integrating real-time web search capabilities to keep AI responses up-to-date and contextually relevant.

#### How does Tavily improve it?
- Tavily enables live searches to retrieve the latest news, research papers, and real-world updates.
- LLMs have a knowledge cutoff (e.g., ChatGPT may not know recent events). Tavily enables AI to bridge this gap by fetching real-time data from the web.

So, let's set up the API key for using this in our project.
It's FREE!!!!!!

### TAVILY SETUP

- Go to Tavily's web page - https://tavily.com/
- Click on Sign Up. You can create a new account or sign up using your Google/Github account.
- Once you successfully sign up, you will then be redirected to the Overview Dashboard as shown below. Your API key will be visible there as shown below (You get 1000 Free API Credits to use)


In [ ]:
# Setting up the Tavily API Key to do web search
os.environ["TAVILY_API_KEY"] = getpass.getpass()

In [ ]:
# import the 'tool' decorator
from langchain.tools import tool
from langchain.tools.retriever import create_retriever_tool

# Define a tool for Amazon product search using @tool decorator
@tool
def amazon_product_search(query: str):
    """Search for information about Amazon products.
    For any questions related to Amazon products, this tool must be used."""

    # Create the retriever tool
    retriever_tool = create_retriever_tool(
        retriever,  # The retriever object that fetches Amazon product data
        name="amazon_search",
        description="Search for information about Amazon products."
    )

    # Execute search based on query
    return retriever_tool.invoke(query)

### Code Explanation:

- Importing `create_retriever_tool`:

`create_retriever_tool` is a utility function in LangChain that allows you to wrap a retriever (a component that fetches relevant information from a data source) into a tool that can be used within an AI workflow.

- Creating the `retriever_tool`:
    - **retriever:** This is a retriever instance (coded in the previous class).
    - **name="amazon_search":** Defines the tool's name, which helps identify it when used in multi-tool environments.
    - **description:** Provides a natural language description of what the tool does, which helps AI models understand when to invoke it.

In [ ]:
# Import necessary classes from LangChain for Tavily integration
from langchain_community.tools.tavily_search import TavilySearchResults

@tool
def search_tavily(query: str):
    """
    Executes a web search using the TavilySearchResults tool.

    Parameters:
        query (str): The search query entered by the user.

    Returns:
        list: A list of search results containing answers, raw content, and images.
    """
    # Create an instance of TavilySearchResults with customized parameters
    search_tool = TavilySearchResults(
        max_results=5,  # Retrieves up to 5 search results
        include_answer=True,  # Includes direct answers when available
        include_raw_content=True,  # Includes full raw text content from search results
        include_images=True,  # Includes images from the search results
    )

    # Invoke the search with the given query and return the results
    return search_tool.invoke(query)



### Code Explanation:

`search_tavily(query: str)`:
This function initializes TavilySearchResults with the following parameters & returns the results for the user query:

    - `max_results=5`: Limits the number of search results to 5.
    - `include_answer=True`: Retrieves direct answers when available.
    - `include_raw_content=True`: Fetches full raw text content from the search results.
    - `include_images=True`: Includes images found in the search results.


### Example:
If the chatbot needs real-time information (e.g., latest stock prices, news, or research articles), it can call TavilySearch to retrieve fresh data from the web.
**Example Query:**
- User: "What are the latest AI trends in 2025?"
- AI (using TavilySearch) → Fetches 5 recent web search results and summarizes them in the response.



In [ ]:
# hwchase17/react is a prompt template designed for ReAct-style
# conversational agents.
prompt = hub.pull("hwchase17/react")

In [ ]:
print(prompt.template)

### Code Explanation:
- **Pulling a Prompt from LangChain Hub (hub.pull())**
LangChain Hub provides predefined prompt templates that can be used in AI applications.
    - The "hwchase17/react-chat" prompt is a ReAct (Reasoning + Acting) style prompt optimized for conversational AI.
    - The pulled prompt will define how an AI agent should structure responses.

**Example scenario:**
    - User: "What’s the latest AI research?"
    - AI: (Fetches real-time research using Tavily and provides an answer.)


In [ ]:
## Create a list of tools: retriever_tool and search_tool
tools = [search_tavily, amazon_product_search]

In [ ]:
react_agent = create_react_agent(
    llm=llm,  # The OpenAI model responsible for reasoning and response generation.
    tools=tools,  # A list of external tools (e.g., web search, product retrieval).
    prompt=prompt  # The ReAct-style prompt guiding the agent's thought process.
)

In [ ]:
# executes the logical steps we created
agent_executor = AgentExecutor(
    agent=react_agent,
    tools=tools,
    verbose=True,
    handle_parsing_errors=True,
    max_iterations = 5 # useful when agent is stuck in a loop
)

### Code Explanation

1. Initialize the OpenAI Language Model (LLM):
 - OpenAI's completion-based model is used to generate responses.
 - Setting `temperature=0` ensures deterministic and factual responses.
 - A lower temperature reduces randomness, making responses more structured.

2. Creating the ReAct Agent:
 - The agent follows the ReAct framework (Reasoning + Acting).
 - It first thinks through the query and then decides whether to call a tool.
 - If a tool is required, it invokes the appropriate one before responding.

3. Initializing the Agent Executor:
 - The AgentExecutor handles query execution and tool invocation.
 - It decides when the agent should call a tool vs. answer directly.
 - The verbose mode provides detailed logs to track agent reasoning and tool interactions.


### Let's try some examples

#### Questions that use Amazon tool only

In [ ]:
res = agent_executor.invoke({"input": "What is the best shoes I can find on Amazon?"})

In [ ]:
print(res['output'])

#### Questions that use the search tool only

In [ ]:
res = agent_executor.invoke({"input": "What is the current weather in Toronto?"})

In [ ]:
print(res['output'])

#### Questions that use both tools

In [ ]:
res = agent_executor.invoke({"input": "How to install Backsplash Wallpaper? Also find some brands on Amazon"})

In [ ]:
print(res['output'])

## Adding Memory and Multi-turn Capabilities

### What is a multi-turn conversation?
<center><img src="https://poly.ai/wp-content/uploads/2020/12/Blog-post-graphics_Multi-1.png"/></center>

A multi-turn conversation is an interactive dialogue where the AI retains context across multiple exchanges, allowing for a natural back-and-forth flow instead of handling each query in isolation.

Example:
User: "What's the weather like in New York?"
AI: "It's 10°C and cloudy."
User: "What about tomorrow?"
AI: "Tomorrow, it will be 12°C with light rain."

Here, the AI remembers "New York" from the first query, maintaining context instead of requiring the user to repeat it.


A multi-turn conversation relies on **memory** to retain context across exchanges, enabling AI to respond meaningfully without requiring repeated inputs.

Adding memory component to the application has several advantages as follows:

1. **Context Retention** – Enables natural and coherent conversations by remembering past interactions.  
2. **Personalization** – Adapts responses based on user history and preferences for a better experience.  
3. **Multi-Turn Conversations** – Supports complex queries and step-by-step interactions without losing track.  
4. **Efficiency & Reduced Repetition** – Eliminates the need for users to repeat information, improving usability.  
5. **Enhanced User Satisfaction** – Creates a more human-like, engaging, and helpful chatbot experience.

In [ ]:
# Creating a Prompt Template

# Pull a predefined prompt from LangChain Hub
# "hwchase17/react-chat" is a prompt template designed for ReAct-style conversational agents.
prompt = hub.pull("hwchase17/react-chat")

# Initialize a chat message history object with a session ID
# This stores the conversation history for a given session, allowing stateful interactions.
memory = ChatMessageHistory(session_id="test-session")
memory.clear()

In [ ]:
# Initialize an Agent with Chat History to maintain conversation context
agent_with_chat_history = RunnableWithMessageHistory(
    agent_executor,  # The main AgentExecutor responsible for executing queries and invoking tools.

    lambda session_id: memory,

    input_messages_key="input",  # Specifies the key in the input dictionary where the user query is stored.

    history_messages_key="chat_history",  # Specifies the key under which the conversation history is stored.
)

### Code Explanation:

- Initializing an Agent with chat history:
 - `session_id` is useful in real-world applications where session tracking is required. Here, it simply returns the in-memory chat history for simplicity.
 - `RunnableWithMessageHistory` ensures the agent retains context across multiple interactions. This is useful for multi-turn conversations where past messages influence future responses.


In [ ]:
# Invoke the agent with chat history
# Let's try a query based on recent data and check whether 'Tavily Search' is triggered
agent_with_chat_history.invoke(
    {"input": "What is the latest fashion trends in 2025?"},
    config={"configurable": {"session_id": "<foo>"}},
)


In [ ]:
# Invoke the agent with chat history
# Let's try a query based on amazon data & check if 'amazon_search' tool is triggered
agent_with_chat_history.invoke(
    {"input": "Can I get these trends from Amazon?"},
    config={"configurable": {"session_id": "<foo>"}},
)


### What Next?

- Try a gradio UI for the optimized version of your chatbot.
- Experiment with various parameters and understand how the output varies!

<hr>